### PR TITLE
CORTX-30863: fix: Handled the conf save invoke at caller end instead …

### PR DESCRIPTION
**Pull request by kailashyadav11**
…of conf_store

Signed-off-by: Kailash Yadav <kailash.yadav@seagate.com>

# Problem Statement
-  Fix the problem where num_xxx keys for the list items are not being saved to the gconf.

# Design
-  [For Bug, Describe the fix here.](https://jts.seagate.com/browse/CORTX-30863)
-  Caller of the add_num_keys API should also explicitly call to save the confstore keys to Gconf.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-   [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-   [x] Unit and System Tests are added
-   [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
-   [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-   [ ] Interface change (if any) are documented
-   [x] Side effects on other features (deployment/upgrade)
-   [x] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
-   [x] JIRA number/GitHub Issue added to PR
-   [x] PR is self reviewed
-   [x] Jira and state/status is updated and JIRA is updated with PR link
-   [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
-   [x] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by tanujashinde0405 — APPROVED** [review]

**Review by nitin-seagate — APPROVED** [review]
LGTM

File: src/provisioner/provisioner.py [hunk 1]
```diff
@@ -127,7 +127,7 @@ def config_apply(solution_config_url: str, cortx_conf_url: str = None,
                     Conf.set(CortxProvisioner._solution_index, key, val.decode('utf-8'))
             CortxProvisioner.apply_cortx_config(cortx_conf, CortxProvisioner.cortx_release)
             # Adding array count key in conf
-            cortx_conf.add_num_keys()
+            CortxProvisioner._add_num_keys(cortx_conf)
```
Comment (ujjwalpl) [inline src/provisioner/provisioner.py:130]: For two lines its worth to put them here for readability and fact that explicit save is called. 
```
        cortx_conf.add_num_keys()
        Conf.save(cortx_conf._conf_idx)
```

File: test/setup/test_provisioner.py [hunk 2]
```diff
@@ -56,6 +65,25 @@ def test_config_apply_bootstrap(self):
...
+        config_path = os.path.join(working_directory,"/sample_config.yaml")
+        with open(config_path, 'w+') as config:
+            config.write(yaml.dump(data))
+        config_url = f"yaml://{config_path}"
+        cortx_conf = MappedConf(config_url)
+        CortxProvisioner._add_num_keys(cortx_conf)
+        Conf.load(test_index, config_url)
+        self.assertEqual(3, Conf.get(test_index, 'num_a'))
+        self.assertEqual(2, Conf.get(test_index, 'c[0]>num_5'))
+        delete_file(config_path)
```
Comment (ujjwalpl) [inline test/setup/test_provisioner.py:86]: Also add a test where
- load config
- add num keys (dont save)
- load config into another index
- num_keys must not be present

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: src/provisioner/provisioner.py (hunk 1 above)
Comment (kailashyadav11) [reply src/provisioner/provisioner.py:130]: @ujjwalpl  correct, but for unit test convenience I have separated the concern of adding num_xxx keys to a different function. Please confirm if you still want this change.

File: test/setup/test_provisioner.py (hunk 2 above)
Comment (kailashyadav11) [reply test/setup/test_provisioner.py:87]: done

File: src/provisioner/provisioner.py (hunk 1 above)
Comment (kailashyadav11) [reply src/provisioner/provisioner.py]: done

File: test/setup/test_provisioner.py [hunk 3]
```diff
@@ -20,13 +20,14 @@
 import os
 import traceback
 import unittest
+import yaml
 from cortx.provisioner.provisioner import CortxProvisioner
+from cortx.utils.conf_store import MappedConf, Conf
```
Comment (ujjwalpl) [inline test/setup/test_provisioner.py:25]: This can be removed if not used in this code.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: test/setup/test_provisioner.py (hunk 3 above)
Comment (kailashyadav11) [reply test/setup/test_provisioner.py]: done

**Review by ujjwalpl — APPROVED** [review]

**Comment by kailashyadav11** [comment]
Got 3 approvals to merge

**Pull request merged**